hicexplorer/hicAggregateContacts.py

```diff
@@ -187,14 +218,193 @@ def read_bed_per_chrom(fh):
+    """
+
+    chrom1, start1, end1 = interval[0]
+    chrom2, start2, end2 = interval[1]
+    if chrom1 not in agg_info["chrom_coord"]:
+        return
+    if chrom2 not in agg_info["chrom_coord"]:
+        return
+    if (int(end1) > agg_info["chrom_coord"][chrom1][1]) or (int(end2) > agg_info["chrom_coord"][chrom2][1]):  # TODO these intervals may still partially be overlapped, shall we keep them?
+        return
+    if (int(start1) < agg_info["chrom_coord"][chrom1][0]) or (int(start2) < agg_info["chrom_coord"][chrom2][0]):
+        return
+
+    bin_id1 = ma.getRegionBinRange(toString(chrom1), start1, end1)
+    bin_id2 = ma.getRegionBinRange(toString(chrom2), start2, end2)
+    if bin_id1 == bin_id2:
+        return
+    if (bin_id1 is None) or (bin_id2 is None):
+        return
+    else:  # If the regions size is bigger than a bin then:
+        bin_id1 = bin_id1[0]
+        bin_id2 = bin_id2[0]
+        if largeRegionsOperation == 'last':
+            bin_id1 = bin_id1[-1]
+            bin_id2 = bin_id2[-1]
+        elif largeRegionsOperation == 'center':
+            bin_id1 = bin_id1[np.floor(len(bin_id1) / 2)]
+            bin_id2 = bin_id2[np.floor(len(bin_id2) / 2)]
+        elif largeRegionsOperation == 'sum':
+            bin_id1 = np.sum(bin_id1)
+            bin_id2 = np.sum(bin_id2)
+        elif largeRegionsOperation == 'mean':
+            bin_id1 = np.mean(bin_id1)
+            bin_id2 = np.mean(bin_id2)
+        elif largeRegionsOperation == 'median':
+            bin_id1 = np.median(bin_id1)
+            bin_id2 = np.median(bin_id2)
+    if bin_id1 > bin_id2:
+        if chrom1 == chrom2:
+            bin_id1, bin_id2 = sorted([bin_id1, bin_id2])
+        else:
+            chr1, str1, en1, bin1 = chrom1, start1, end1, bin_id1
+            chrom1, start1, end1, bin_id1 = chrom2, start2, end2, bin_id2
+            chrom2, start2, end2, bin_id2 = chr1, str1, en1, bin1
+    agg_info["counter"] += 1
+    if agg_info["counter"] % 50000 == 0:
+        log.info("Number of contacts considered: {:,}".format(agg_info["counter"]))
+
+    bin_size = ma.getBinSize()
+    chrom1_bin_range = ma.getChrBinRange(toString(chrom1))
+    chrom2_bin_range = ma.getChrBinRange(toString(chrom2))
+
+    if chrom1 == chrom2:  # chrom1 == chrom2 can happen in intra or all
+        if (chrom1 not in agg_info["agg_total"]) and (mode == "intra-chr") and (perChr == True):
+            agg_info["agg_total"][chrom1] = 0
+            agg_info["agg_matrix"][chrom1] = []
+            agg_info["agg_diagonals"][chrom1] = []
+            agg_info["agg_contact_position"][chrom1] = []
+            agg_info["agg_center_values"][chrom1] = []
+
+        min_dist, max_dist = range.split(":")
+        log.info("checking range {}-{}".format(min_dist, max_dist))
+        assert int(min_dist) < int(max_dist), "Error lower range larger than upper range"
+        min_dist_in_bins = int(min_dist) // bin_size
+        max_dist_in_bins = int(max_dist) // bin_size
```

Comment:
Could all this be done only once and stored in `agg_info` instead of being evaluated for each interval pair?